frontend/src/app/shared/components/tree/tree.component.ts

```diff
@@ -0,0 +1,107 @@
+import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
+
+import { TreeNode } from './models/TreeNode.model';
+
+@Component({
+    selector: 'app-tree',
+    templateUrl: './tree.component.html',
+    styleUrls: ['./tree.component.sass'],
+})
+export class TreeComponent implements OnInit {
+    @Input() asCheckList: boolean = false;
+
+    @Output() selectionChange = new EventEmitter<{ selectedNodes: TreeNode[]; originalStructure: TreeNode[] }>();
+
+    @Input() treeData: TreeNode[] = [
+        {
+            name: 'Category 1',
+            children: [{ name: 'Subcategory 1.1' }, { name: 'Subcategory 1.2' }],
+        },
+        {
+            name: 'Category 2',
+            children: [{ name: 'Subcategory 2.1' }, { name: 'Subcategory 2.2' }],
+        },
+    ];
+
+    public filteredTreeData: TreeNode[] = [];
+
+    ngOnInit(): void {
+        this.filteredTreeData = this.treeData;
+    }
+
+    public filterTree(event: Event): void {
+        const searchTerm = (event.target as HTMLInputElement).value;
+
+        if (!searchTerm) {
+            this.filteredTreeData = this.treeData;
+
+            return;
+        }
+
+        const filteredData: TreeNode[] = [];
+
+        this.treeData.forEach((category) => {
+            const matchingSubcategories = category.children?.filter((subcategory) => {
+                return subcategory.name.toLowerCase().includes(searchTerm.toLowerCase());
+            });
+
+            if (matchingSubcategories?.length) {
+                filteredData.push({
+                    name: category.name,
+                    children: matchingSubcategories,
+                });
+            }
+        });
+
+        this.filteredTreeData = filteredData;
+    }
+
+    public toggleSelect(node: TreeNode): void {
+        if (node.children) {
+            const allChildrenSelected = node.children.every((child) => child.selected);
+
+            node.selected = !allChildrenSelected;
+            node.children.forEach((child) => (child.selected = node.selected));
```

Comment:
Let's not use foreach, map please

Reply:
should I use map even when I don't transform the array elements?

Reply:
Yep, you change their values, it's better to not use mutable structures